### PR TITLE
New module: management of the Nuage Networks VSP SDN solution (network/nuage/nuage_vspk)

### PR DESCRIPTION
**Pull request by pdellaert**
##### SUMMARY
Adding a new module to manage the Nuage Networks VSP SDN solution through its Python SDK.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
nuage_vspk

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 17 2016, 23:03:43) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

### CODE REVIEW

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: Please delete all occurrences of  `choices: []`, that's the default, so not needed.

This comment applies to a number of places in this module.

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: Please delete all occurrences of  `aliases: []`, that's the default, so not needed.

This comment applies to a number of places in this module.

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: The only `version_added` should be at the top level, the rest can be removed.

This comment applies to a number of places in this module.

This is the **Ansible** version that the option was changed.

Please see https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html for more information

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py:454]: If there are going to be a number of modules that need this logic, it may be good to move this into common code `module_utils` see https://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html for more info.

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py:36]: If some options are required together this can be enforced in the argspec,

See the following draft documentation for examples on this:
https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: I think most if not all of this can be removed and replaced with argspec code

https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py:1016]: https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (ganeshrn) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: Since `children`, `properties` and other's are defined as module args, the key will be present in `params` and initialized to is a default value. Hence the `if` check here and above is not required.
`self.children = module.params['children']` is sufficient

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (pdellaert) [reply lib/ansible/modules/network/nuage/nuage_vspk.py:36]: Done, thanks for the tip!

One remark, the doc stipulates the use of `()` brackets, however, that caused issues in some cases, i've used `[]` everywhere.

Sub options are apparently not available yet in 2.3, as such, i've added them, according to the code in module_utils/basic.py

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (pdellaert) [reply lib/ansible/modules/network/nuage/nuage_vspk.py:454]: This is something we are planning once we move to more modules, we'll split the general functionalities (not only this, but building the connection as well) into a module_util. Currently this is not needed yet, i think.

**Comment by pdellaert** [comment]
@gundalow , @ganeshrn, thank you for the review, comments and especially tips!

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [reply lib/ansible/modules/network/nuage/nuage_vspk.py:454]: Sure, that sounds sensible.

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [reply lib/ansible/modules/network/nuage/nuage_vspk.py:36]: I believe I may have made a mistake in my documentation regarding `()` vs `[]`, my apologies for any confusion caused, it's not been reviewed yet. I'll make sure that gets updated. 

New modules go into `devel` which will become the Ansible 2.4 release. What's the comment regarding 2.3 regarding?

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [inline lib/ansible/modules/network/nuage/nuage_vspk.py]: Now you are using subboptionsn you only need `no_log` on password, & api key.

Comment (pdellaert) [reply lib/ansible/modules/network/nuage/nuage_vspk.py]: @gundalow the comment on 2.3 was that i was testing with 2.3. Which makes me realise that i should be testing directly with the devel branch instead. 

I'll have a look and also change the no log to just the password and key (potentially certificate as well, it's not that crucial, but they are stored typically together and both point to files)

Comment (pdellaert) [reply lib/ansible/modules/network/nuage/nuage_vspk.py]: @gundalow while doing negative tests, i ran into an issue in the current `module_utils/basic.py` code, namely:
https://github.com/ansible/ansible/blob/757758cd1a09beb7b99aef7b5dab04bb2f4fc428/lib/ansible/module_utils/basic.py#L1739
```
value = self.params[k]
```

This will fail on when doing the check of the sub options, with the following traceback (cause: it is trying to find the sub option in the main params instead of the sub params):
```
Traceback (most recent call last):
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_module_nuage_vspk.py", line 1052, in <module>
    main()
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_module_nuage_vspk.py", line 1041, in main
    supports_check_mode=True
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 759, in __init__
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 1759, in _check_argument_types
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 1739, in _check_argument_types
KeyError: 'api_enterprise'
```

The code should be:
```
value = param[k]
```

(In case you are interested, my negative test cases can be found here: https://github.com/nuagenetworks/vspk-ansible/blob/beccbf3269a3dd5163422814506e9a4b5b244240/roles/nuage-vspk-negative-tests/tasks/main.yml )

Comment (pdellaert) [reply lib/ansible/modules/network/nuage/nuage_vspk.py]: @gundalow And i also noticed that if i remove the no_log from the auth section and move it to the sub options, it actually will print those values in debug output:

```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "auth": {
                "api_enterprise": "csp",
                "api_password": "csproot",
                "api_url": "https://localhost:8443",
                "api_version": "v4_0"
            },
            "command": "wait_for_job",
            "id": "dumy-job-uuid",
            "type": "Job"
        }
    },
    "msg": "missing required arguments: api_username"
}
```

**Comment by pdellaert** [comment]
@gundalow I've made the changes as suggested, however, running this will fail because of the bug mentioned here: https://github.com/ansible/ansible/issues/25096

For the no_log of sub options, i created issue #25097

Feel free to run the negative tests, you don't need any Nuage component to do so: 
https://github.com/nuagenetworks/vspk-ansible/blob/beccbf3269a3dd5163422814506e9a4b5b244240/roles/nuage-vspk-negative-tests/tasks/main.yml

File: lib/ansible/modules/network/nuage/nuage_vspk.py
Comment (gundalow) [reply lib/ansible/modules/network/nuage/nuage_vspk.py]: 1) Can you please raise a bug (or PR if you know the fix) for the `basic.py` issue
2) suboptions don't (currently) inherit `no_log`, or `type`
So you will need to set that explicitly on the sub options as needed

Comment (gundalow) [reply lib/ansible/modules/network/nuage/nuage_vspk.py]: Looks like #2 has been addressed in https://github.com/ansible/ansible/pull/24895/commits/78c26d9ff6bf80d7967196beae876fbfee6c2d9f

**Comment by gundalow** [comment]
@pdellaert Looks like #25096 has been fixed in `devel`

**Comment by gundalow** [comment]
All network modules added into 2.4 need to include Unit Tests. So that will need adding before this can be merged. 

Have a look at 
* `test/units/modules/network/`
* `test/units/module_utils/` # For when you start adding module_utils code
* https://docs.ansible.com/ansible/dev_guide/testing_units.html

How do the tests in https://github.com/nuagenetworks/vspk-ansible/blob/beccbf3269a3dd5163422814506e9a4b5b244240/roles/nuage-vspk-negative-tests/tasks/main.yml work without having a Nuage device? I see them talking to a web endpoint with credentials.

I wonder if we could add them into our CI testing.

**Comment by pdellaert** [comment]
@gundalow For the negative tests, you do not need any system to run, these just verify that all required inputs are met without creating a connection to the Nuage VSP environment.

For any other tests, a setup is required. A potential solution could be to use our https://nuagex.io service. If you create an account, please let me know so i can get it activated as soon as possible. 

I'll look into the unit tests, we do differ from the rest of the network modules with unit tests, it seems. As these modules use CLI commands and output from their respective platforms and we talk to our HTTPS REST API to manage the Nuage solution. Unit tests are still very useful, of course, we already have unit tests in the libraries used by the module, i'll add tests for the module specific.

**Comment by pdellaert** [comment]
Seems like something went wrong with my rebase... i'll create a new branch and push that

**Comment by pdellaert** [comment]
@gundalow Sorry about the mess with the rebase, my git expertise is not where it needs to be it seems.

I've cleaned up, but the labels are still applied, which are unnecessarily there, i think.

I've added unit tests, which cover 91% of the code. The only things that are not covered are check mode. 

i've also removed certain code from the module which proved unreachable during testing, as they were redundant tests.

On the failure of the unit tests: Our packages (vspk/bambou) currently do not support Python3 yet, i hope that is not a blocker for this (as adding that will take some effort). 

I'm adding a conditional check

File: test/units/modules/network/nuage/nuage_module.py
Comment (gundalow) [inline test/units/modules/network/nuage/nuage_module.py]: Please wrap the imports in a try as shown in https://github.com/ansible/ansible/pull/25762/files

This will help people that are running tests locally (ie outside of Shippable) that don't for whatever reason want to have all the `requirements/units.txt` libraries installed

File: test/runner/requirements/units.txt
Comment (gundalow) [inline test/runner/requirements/units.txt]: `# requirements for network/nuage specific modules`

**Comment by gundalow** [comment]
Once the above is resolved we can get this merged, CHANGELOG updated and you listed as maintainers

**Comment by pdellaert** [comment]
@gundalow Made the changes, should be good. Thank you for your support!

**Comment by gundalow** [comment]
@pdellaert We've been discussing this internally and we believe if you mock the imports (and remove the file from `requirements/units.txt` this will be better.

Have a look at this commit https://github.com/ansible/ansible/pull/25450/commits/535a2b52682de0bbe84d398b4d552dc80219fee7
from this PR https://github.com/ansible/ansible/pull/25450 for an example.

It's on the Core Team's todo list to add some documentation in for this

**Comment by pdellaert** [comment]
@gundalow mocking the VSPK and bambou module in its whole might be complicated on doing, it is a pretty complex integration.

Is that what you are asking? If so, i'll probably require a while...

**Comment by gundalow** [comment]
@pdellaert That's correct.
If you available for the Testing Working Group meeting today we will be discussing mocking https://github.com/ansible/community/issues/110

**Comment by pdellaert** [comment]
@gundalow I'll join, my nickname is 'straks' (just to confuse every dutch person in the room ;))

**Comment by gundalow** [comment]
@pdellaert Hi, just wondering how you were getting on with the integration tests for this?

**Comment by pdellaert** [comment]
@gundalow i have it on my todo list for the next coming days/week (depending how much time i can spend on this over the weekend)

**Comment by pdellaert** [comment]
@gundalow I've created a bunch of Integration tests. If you want to go over this, we can discuss on IRC.

If you need me to close this PR and open a clean one, let me know.

(PS, getting the hang of rebase, sorry for all the messages)

**Comment by dagwieers** [comment]
-label vmware

**Review by gundalow — APPROVED** [review]

**Pull request merged**
